File: Element/LIS.cpp
// Longest increasing subsequence - O(n^2) complexity

#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
    static int LIS(vector<int> &arr){
        int n = arr.size();
        int dp[n];
        for(auto i = 0; i < n; ++i){
            dp[i]=1;
            for(auto j = 0; j < i; ++j){
                if(arr[j] < arr[i])
                    dp[i] = max(dp[i],dp[j]+1);
            }
        }
        return *max_element(dp,dp+n);
    }
};
int main(){
    vector<int> test = {10, 22, 9, 33, 21, 50, 41, 60};
    printf("%d",Solution::LIS(test));
}

File: Element/LIS.py
# Longest Increasing Subsequence DP - O(N²) complexity

class Solution(object):
    def LIS(self,arr)->int:
        self.n = len(arr)
        self.DP = [None for i in range(self.n)]
        for i in range(self.n):
            self.DP[i] = 1
            for j in range(i):
                if(arr[j] < arr[i]):
                    self.DP[i] = max(self.DP[i],self.DP[j]+1)
        return max(self.DP)


if __name__=='__main__':
    test = [1,7,3,5,9,4,8]
    s = Solution()
    print(s.LIS(test)) # 4 (1,3,5,8)
